[Merge] Merge Integration of frontend and backend

nonohuang0819 opened the pull request:
# [Merge]前後端 API 並完成資料交互

## 背景  
完成了前後端整合，確保前端能夠正確地與後端 API 進行資料交互

## 主要變更  

### 1. **API 整合**  
- 完成前端對後端 API 的串接，包含資料請求 (GET)、提交 (POST)、更新 (PUT) 及刪除 (DELETE) 功能。    

### 2. **資料格式處理**  
- 確保前端發出的請求與後端需求的資料格式一致。  
- 新增 Error Handling 機制，處理 API 請求的例外情況（如 404、500 等）。  


## 影響範圍  
- **後端：**  
  - User, Folder, Quiz, Question 更新create 有更新路逕，請注意backend route的網址  

- **前端：**  
  - homepage
     - signup page: signup 完不會自己登入，會等待兩秒自動轉導到homepage，需要自己進行登入
     - login page: signup 完，會轉導到homepage
     - 實做了GetRandQuiz function，可以隨機抽取想要的數量到homepage裡
  - Userlibrary page: 
     - 能做Folder 的CRUD
     - 能做Quiz 的CRUD
     - 能做Question 的CRUD
  - API Client
     - 修改VITE
       
## 檔案結構
backend
├── src
│   ├── controllers
│   │   ├── Folder
│   │   │   └── FolderController.js
│   │   ├── Quiz
│   │   │   └── QuizController.js
│   │   ├── Question
│   │   │   └── SO_QuestionController.js
│   │   └── User
│   │       └── UserController.js
│   ├── db_services
│   │   ├── Folder
│   │   │   └── FolderServices.js
│   │   ├── Quiz
│   │   │   └── QuizServices.js
│   │   ├── Question
│   │   │   └── SO_QuestionServices.js
│   │   └── User
│   │       └── UserServices.js
│   ├── routes
│   │   ├── Folder
│   │   │   └── FolderRoutes.js
│   │   ├── Quiz
│   │   │   └── QuizRoutes.js
│   │   ├── Question
│   │   │   └── SO_QuestionRoutes.js
│   │   └── User
│   │       └── UserRoutes.js
│   └── database.js

frontend
│   ├── components
│   │   ├── Home
│   │   │   └── SideBar.vue
│   │   ├── NavBar.vue
│   │   ├── UserLibrary
│   │   │   ├── DisplayQuestion.vue
│   │   │   ├── EditQuiz
│   │   │   │   ├── QuestionCreatePop.vue
│   │   │   │   └── QuestionEditPop.vue
│   │   │   └── Library
│   │   │       ├── FolderCreatePop.vue
│   │   │       ├── FolderEditPop.vue
│   │   │       └── LeftBar.vue
│   ├── main.js
│   ├── router
│   │   └── index.js
│   ├── service
│   │   └── LibraryApi
│   │       ├── ApiClient.js
│   │       ├── FolderAPI.js
│   │       ├── QuestionAPI.js
│   │       └── QuizAPI.js
│   ├── stores
│   │   └── Userlibrary
│   │       └── QuizQuestionStore.js
│   └── views
│       ├── Home
│       │   └── HomePage.vue
│       ├── Signup
│       │   └── SignupPage.vue
│       ├── Login
│       │   └── LoginPage.vue
│       └── UserLibraryPages
│           ├── EditQuiz.vue
│           └── UserLibrary.vue

## 相依性
- views 底下的` homepage.vue ` 使用了GetRandQuiz API ，可透過改number隨機抽取想要的quiz數量到homepage裡
-  account API 
    - component/SignView 底下的 login.vue call login function signup
    - component/SignView 底下的 signup.vue call signip function
    - component 底下的Navbr.vue call login 所創的userdata 這樣才能顯示登入的使用者名字
- Folder API 
    -  component/Userlibrary/FolderCreatePop.vue call 
    - component/Userlibrary/FolderEditPop.vue call
    - component/UserlibraryLeftBar.vue call
- Quiz API 與 QuizEditBlock.vue, FolderEditPop.vue, HomePage.vue, UserLibrary.vue
- Question API 與  QuestionCreatePop.vue,  QuestionEditPop.vue and UserLibrary.vue

## 測試結果  
- **API 測試：** 所有關鍵功能的 API 測試已通過。  

## 使用方式
- 請記得改前端的.env ，並改成  VITE_API_BASE_URL = 'http://localhost:3000' 
- `git switch integration_basic_CRUD`
- 打開兩個terminal
- 一個`cd backend` ，另一個`cd frontend` 
- 都做 ` npm i `
- 都做 ` npm run dev `
- 在瀏覽器打開前端的網址

## 小Bug ，但不影響使用者體驗

(1)
若點edit quiz，在重新整理
![image](https://github.com/user-attachments/assets/9fc72cbf-1d8b-4a73-9cd8-d6eab2db86d4)
在這頁面重新整理
![image](https://github.com/user-attachments/assets/aa54066d-3635-44aa-824e-c02a7b2efe8d)
會變成這樣
![image](https://github.com/user-attachments/assets/8038c5f5-ecc1-4be8-a3b3-a7c75ad62e7e)

(2)
若點create question，在重新整理
![image](https://github.com/user-attachments/assets/b3bafcf9-8699-49c1-a6a0-cecdd24c4355)
在這頁面重新整理
![image](https://github.com/user-attachments/assets/24a60306-9ad3-4ae0-8c5a-b66033927e31)
會變成這樣
![image](https://github.com/user-attachments/assets/b876d970-f99d-45f5-8e07-07f5af525527)

bscny committed:
[Style][Fix] fix some bugs and restyle a bit

bscny committed:
[Create][Fix] adjust service folder and build up an axios workflow

bscny committed:
[Chore] delete redundant files

Commit:
[Create]build User login confirmation function

bscny committed:
[Restruct][Fix] make quiz grid only focus on site rendering, move fetch data to parent section

Commit:
[Create]Build User login confirmation function

Commit:
[Create]Build random Quiz selection

Commit:
[Create]Build random Quiz selection

bscny committed:
Merge pull request #14 from bscny/hotfix_quizGrid

[Merge] merging hotfix-quizGrid to integration branch

Commit:
[Fix] fix backend and front end

Commit:
[Fix]ApiClient.js API_BASE_URL change to VITE_API_BASE_URL

Commit:
[Create] connect frontend sign up function and backend createuser function

Commit:
[Create] connect frontend sign up function and backend createuser function

Commit:
[Create] connect frontend userlogin function and backend userconfirmation function

Commit:
[Create] signup 完，過兩秒會轉導到homepage，並且不會有登入的狀態

Commit:
[FIX] connect API can get specific user's folder

Commit:
[FIX] connect API to create folder

Commit:
[FIX] connect API create/ update and delete folder

Commit:
[Update]Update Seed

Commit:
[Nothing]

Commit:
[Delete]delete node-module

Commit:
[FIX]connect API to show quiz and its question

Commit:
[FIX]connect quiz API to create, show, update  and delete

Commit:
[UPDATE] update seed.sql

Commit:
[Create]integration basicCRUD All done

Commit:
[FIX]connet API to use GetRandQuiz function

Commit:
[Nothing]

Commit:
[Change] comment change

bscny commented:
.env.example 沒有更新道喔，然後 [nothing] 是啥鬼 lmaooooo

nonohuang0819 commented:
我來改

Commit:
[Update] .env.example update  and some comment

nonohuang0819 commented:
修復成功

jingxiang0405 commented:
我今天熬夜做資訊視覺化
星期二晚上看

bscny commented:
不行啦，看這個不會花你多久時間，你又不是要寫

bscny commented:
然後 @nonohuang0819 你趕快把 `.env.example` 更新好嗎

bscny commented:
你完全沒更新到
![image](https://github.com/user-attachments/assets/565c29e7-180d-4c88-8227-4dbbd7acf5ce)

nonohuang0819 commented:
我更了呀

bscny commented:
沒有，我這裡抓下來也沒看到有更

Commit:
[Update] update .env.example

nonohuang0819 commented:
再看一次
這次應該有了

bscny commented:
# 大問題

1. sign up 完跳到主畫面沒用，因為你沒設定 localstorage
2. 請看下圖
![image](https://github.com/user-attachments/assets/ac739c06-6a66-4e78-89cd-c77fc1110f5d)
需要按完 `F5` 之後才會
![image](https://github.com/user-attachments/assets/b7b342a1-c81a-41f6-894a-9c8e75269406)
3. 你這裡根本沒打到 server
![image](https://github.com/user-attachments/assets/6fbfdf13-c1d4-4395-a310-5146332b46a5)
4. >小Bug ，但不影響使用者體驗
    - 這就代表你 views 裡面 userLibrary 的 create 有東西沒處理好

# 小問題

1. sign up 完跳到主畫面為啥要等兩秒
2. 一般設定 localstorage 不會再 service 裡面，會在 views 裡面

nonohuang0819 commented:
我現在來修

Commit:
[Update] 能立即修改folder的name，可以直接看到刪除和創建的結果

bscny commented on project/frontend/src/views/UserLibraryPages/UserLibrary.vue:
```diff
@@ -71,39 +87,56 @@ export default{
+                await updateFolder(editedFolder.Folder_id, editedFolder);
+                for (let i = 0; i < this.folders.length; i++) {
+                    if (this.folders[i].Folder_id == editedFolder.Folder_id) {
+                        this.folders[i] = editedFolder;
```
這裡這樣寫 Vue 可能會偵測不到喔，你試改一下，如果 site render 沒有成功就改成
```js
this.folders= [
    ...this.choseAns.slice(0, this.curDisplaying - 1),
    editedFolder,
    ...this.choseAns.slice(this.curDisplaying),
]
```

bscny replied:
你應該在這裡問喔

nonohuang0819 replied:
Questionpop 沒影響不用管
不確定是不是這樣子
![image](https://github.com/user-attachments/assets/537d6d2d-c69b-466d-8144-7cc7718623fb)

Commit:
[Hotfix]Fix FolderEdited 改成這種作法if (index !== -1) {
                    this.folders = [
                        ...this.folders.slice(0, index),
                        editedFolder,
                        ...this.folders.slice(index + 1),
                    ];
                }

nonohuang0819 commented:
Questionpop 沒影響不用管
不確定是不是這樣子
![image](https://github.com/user-attachments/assets/537d6d2d-c69b-466d-8144-7cc7718623fb)

bscny commented on project/frontend/src/views/UserLibraryPages/UserLibrary.vue:
```diff
@@ -112,10 +112,13 @@ export default{
                 await updateFolder(editedFolder.Folder_id, editedFolder);
-                for (let i = 0; i < this.folders.length; i++) {
-                    if (this.folders[i].Folder_id == editedFolder.Folder_id) {
-                        this.folders[i] = editedFolder;
```
你這裡不就找到 i 了, ==

bscny replied:
何必跑兩次

nonohuang0819 replied:
沒啦
那是修改前的程式碼
實際長這樣
![image](https://github.com/user-attachments/assets/a939c13e-0249-4984-b579-256254518369)

nonohuang0819 commented:
signup完，不能去主畫面嗎?
我是想說signup 成功會出現redirecting home page，它會自動轉導到主畫面，至於那兩秒其實不一定要，我可以刪掉

Commit:
[Update] signup完不用再等2秒轉到homepage

bscny commented:
問題是，你沒去成功阿，你自己試試看，使用者名稱都沒出來，功能也用不了

nonohuang0819 commented:
不是老哥
我沒打算讓他sign up完直接登入
所以她本來就不會有顯示名字
要自己手動登入才可以

還是我搞錯意思

你看
註冊:
![image](https://github.com/user-attachments/assets/f3159f70-07a6-42f0-b2f7-b7345f67beaa)
轉導:
![image](https://github.com/user-attachments/assets/c9f6d55e-54f7-4799-b148-e6d6e0277e7e)
本來就不顯示:
![image](https://github.com/user-attachments/assets/1ad154e1-809b-410d-b4c2-0ff7286daf89)

要自己login 才顯示
![image](https://github.com/user-attachments/assets/085ff887-a4de-471f-948b-63de0e3f25e0)

bscny commented:
對，這裡有很大的問題啊，應概要直接登入成功阿，她都註冊完了為啥還要再登入一次

nonohuang0819 commented:
我想說像我們註冊軟體帳號也是一樣
就會讓使用者註冊完重新登入
這做法不太好嗎?
如果不太好我現在修

bscny commented:
沒關西，你先弄其他的，這是非基本問題

bscny commented:
> 我想說像我們註冊軟體帳號也是一樣 就會讓使用者註冊完重新登入 這做法不太好嗎? 如果不太好我現在修

吐槽一下，那明明都是註冊完之後收信，收完信之後自動跳轉，哪來的重新登入 www

nonohuang0819 commented:
有道理!!! 靠邀但我先解決其他的
辛苦你了!!!嗚嗚嗚嗚嗚嗚

bscny commented:
沒事，我昨田整合到 7 點，不信去看小賴的專案，組長的負擔，而且你要慶幸你不是扛著兩個只會用 chatGPT 的雷包，我直接重寫其中一個人的 part，我星期二有空再跟你們分享
![image](https://github.com/user-attachments/assets/f3c01b4f-4611-4435-bddb-50a09d9360b7)

nonohuang0819 commented:
阿你今天會熬夜嗎 
我跟你dc討論一下

bscny committed:
[Fix][Style] fix some bugs in user inventory

bscny committed:
[Fix] fix all bugs in userInventory

Commit:
[Update]user local storage move to login page

Commit:
[FIX]put local storage to right position

nonohuang0819 commented:
> # 大問題
> 1. sign up 完跳到主畫面沒用，因為你沒設定 localstorage
> 2. 請看下圖
>    ![image](https://private-user-images.githubusercontent.com/115096274/399168007-ac739c06-6a66-4e78-89cd-c77fc1110f5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0ODk5MjksIm5iZiI6MTczNTQ4OTYyOSwicGF0aCI6Ii8xMTUwOTYyNzQvMzk5MTY4MDA3LWFjNzM5YzA2LTZhNjYtNGU3OC04OWNkLWM3N2ZjMTExMGY1ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyOVQxNjI3MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNWU1YTE5NjQwYjFjYjVlM2UzNWRiNmEwNjg5OTI2ZjczNjM0NGFjNDU3OTkzZWJmN2FiMzc5NDkzYmJmZmU3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6B2y9ASGc74fr6POl6bDC1zSGHYucEcuxZrpEpzF250)
>    需要按完 `F5` 之後才會
>    ![image](https://private-user-images.githubusercontent.com/115096274/399168046-b7b342a1-c81a-41f6-894a-9c8e75269406.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0ODk5MjksIm5iZiI6MTczNTQ4OTYyOSwicGF0aCI6Ii8xMTUwOTYyNzQvMzk5MTY4MDQ2LWI3YjM0MmExLWM4MWEtNDFmNi04OTRhLTljOGU3NTI2OTQwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyOVQxNjI3MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMjU5NWY5YjIxMzRkZGYwNDA2MDhiYWZmY2MyYzgxM2JjNjBlMGU5NzJiODM4YzE4OWFjMjllYTY2NzM2NjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JtfCm4Fu5sFsxhEsN15KAKQajhlmKbn-WGtgt-dSf6E)
> 3. 你這裡根本沒打到 server
>    ![image](https://private-user-images.githubusercontent.com/115096274/399168182-6fbfdf13-c1d4-4395-a310-5146332b46a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0ODk5MjksIm5iZiI6MTczNTQ4OTYyOSwicGF0aCI6Ii8xMTUwOTYyNzQvMzk5MTY4MTgyLTZmYmZkZjEzLWMxZDQtNDM5NS1hMzEwLTUxNDYzMzJiNDZhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyOVQxNjI3MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZGRiMDc3NGVkMDM5MWY4Nzg3YTg5YTkwZjRiNTg4MzgwYWY3YWExNTBhNjg4MDczODQ3NWQ0ZWVmZmRmYTJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.HdAupYeZ4tr2fiWk9tkbd363VBnbZunbJJCy_16lxFA)
> 4. > 小Bug ，但不影響使用者體驗
>    
>    
>    
>    * 這就代表你 views 裡面 userLibrary 的 create 有東西沒處理好
> 
> # 小問題
> 1. sign up 完跳到主畫面為啥要等兩秒
> 2. 一般設定 localstorage 不會再 service 裡面，會在 views 裡面

nonohuang0819 commented:
## 問題待處理
### 當創建完新quiz時，無法改名/也無法刪除
![image](https://github.com/user-attachments/assets/e52ffcfb-ae59-499c-bfb6-1ad3582998ae)

### 當創建完新quiz時，無法創建新question
![image](https://github.com/user-attachments/assets/ba123a4a-2bae-414d-ac60-7e7037329967)

nonohuang0819 commented:
## 已修復
將localstorage 放到合適的地方

bscny commented:
昨天試不都可以嗎

bscny commented:
然後 你的 console 好亂，這樣怎麼 debug 啊，我完全沒看到錯誤訊息在哪行欸

nonohuang0819 commented:
你晚點再試試看
我後面又跑出這bug

bscny commented:
不 你也要試著 debug，架構是一定沒問題了

nonohuang0819 commented:
我會再來debug
啥鬼亂 我昨天後面只動login signup而已
其他我都沒動

jingxiang0405 committed:
[Fix] route to /Home when user logs out.

bscny commented:
老歌

bscny commented:
這不是你的分支，不能這樣，萬一 @nonohuang0819 也在寫呢? 你這樣會導致她要 3 way 自己的分支

jingxiang0405 commented:
呃..他不會改到NavBar
我這裡小改一下

bscny commented:
你要馬:

1. 從這條分支開一條新的分支，PR 近來
2. 跟他說你接下來都不要動我要直接改了，跟我前天那樣

bscny commented:
一樣啊，紀錄會有問題，即使妹有 conflict，github 的紀錄會長得很奇怪，比如說，會有一個 commit 寫merge 一串網址

jingxiang0405 commented:
> 你要馬:
> 
> 1. 從這條分支開一條新的分支，PR 近來

可以在這條Merge之前放PR進來嗎

bscny commented:
當然可

jingxiang0405 commented:
> ## 問題待處理
> ### 當創建完新quiz時，無法改名/也無法刪除
> ### 當創建完新quiz時，無法創建新question
@bscny 你的UserLibrary/EditQuiz要不要改成UserLibrary/EditQuiz/<quiz_id>

bscny commented:
不要，這樣不安全，我們小賴那邊就是這樣寫的，不想再錯一次了

jingxiang0405 commented:
> 不要，這樣不安全，我們小賴那邊就是這樣寫的，不想再錯一次了

為啥

Commit:
Merge branch 'main' into integration_basic_CRUD

Commit:
[Fix] create quiz 後找不到quiz id 和 create question 後找不到So_id

nonohuang0819 commented:
簡單解釋:
做法:重新create quiz/ question完 ，立刻fetch data，這樣就可以避免，update / delete undefine 

已修復:
create quiz/ question完，無法立刻修改或是刪除

bscny commented:
沒錯 是不是很簡單啊，幾行就結束了，沒你想的複雜

測試完畢了喔，功能都沒問題

bscny commented:
> 為啥

這樣 quiz_id 不就爆露了，使用者不需要知道 id，他只要名子

nonohuang0819 commented:
否 好難 我在電腦前面看超久都沒進展

nonohuang0819 commented:
認真問 我可以怎麼做 我想了蠻久的 找不的做法

bscny commented:
> 認真問 我可以怎麼做 我想了蠻久的 找不的做法

疑… 我不是在你面前接過相同的東西嗎.. 只是我 Fetch 的是 folder… 事實上你也只改了這些啊，絕對不難
![DE602FE1-E0EA-4F5C-B71B-B0E60CA4B7CC](https://github.com/user-attachments/assets/64306a94-c393-4d00-a31b-c51d49e94fd6)

bscny commented:
@jingxiang0405 不要每次都等你

jingxiang0405 commented:
沒問題

nonohuang0819 commented:
@bscny 游爸爸這樣的話我可以merge嗎 好想享受按下去的快感

bscny commented:
我已經給過了啊，你往上滑

nonohuang0819 commented:
來了
ㄚㄚㄚㄚㄚ
